### PR TITLE
Apply GumTree actions to PSI

### PR DESCRIPTION
**Pull request by nbirillo**

### CODE REVIEW

**Review by elena-lyulina — CHANGES_REQUESTED** [review]
Cool, thanks!

File: src/main/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformer.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            is Insert -> action.apply(nodesObserver)
+            is Update -> action.apply()
+            is Move -> action.apply(nodesObserver)
+            else -> error("Unknown GumTree action ${action.name}")
+        }
+    }
+
+    // TODO: it seems we should use simple <applyAction>
+    fun applyActions(actions: List<Action>) {
+        val nodesObserver = NodesObserver()
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply src/main/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformer.kt:49]: A field for the class? We should have a new instance of the insertedNodesIds for each call `applyAction` function, isn't it?

File: src/test/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformerTest.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        val dstPsi = myFixture.getPsiFile(dstFile)
+        val expectedCode = dstPsi.text
+        val srcContext = Util.getTreeContext(srcPsi, numbering)
+        val dstContext = Util.getTreeContext(dstPsi, numbering)
+        val matcher = Matcher(srcContext, dstContext)
+        val actions = matcher.getEditActions()
+        val w = PsiElementTransformer(project, srcPsi, dstPsi, numbering)
+        WriteCommandAction.runWriteCommandAction(project) {
+            w.applyActions(actions)
+        }
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply src/test/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformerTest.kt:72]: I thought that we would like to add an intermediate stage in which we will reorder the actions and select only N of them, so I separated these functions

File: src/test/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformerTest.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    var dstFile: File? = null
+
+    @Test
+    fun `apply src to dst actions`() = convertSrcToDst(srcFile!!, dstFile!!)
+
+    @Test
+    fun `apply dst to src actions`() = convertSrcToDst(dstFile!!, srcFile!!)
+
+    // TODO: In the <fail> folder cases are stored which are incorrect even with GumTree trees
+    //  and GumTree internal actions
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply src/test/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformerTest.kt:45]: Oh sorry, I forgot to push it. I found only one example with this problem

File: src/main/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformer.kt [hunk 4]
```diff
@@ -11,45 +11,48 @@ import org.jetbrains.research.ml.ast.gumtree.tree.Numbering
...
-    srcPsi: PsiElement,
-    dstPsi: PsiElement,
-    private val numbering: Numbering
+data class PsiTransformation(
+    private val srcPsi: PsiElement,
+    private val dstPsi: PsiElement,
+    private val numbering: Numbering,
+    var srcPsiNodes: MutableList<PsiElement> = ArrayList(),
+    var dstPsiNodes: MutableList<PsiElement> = ArrayList(),
+    val insertedNodesIds: MutableSet<Int> = HashSet()
```
Comment (elena-lyulina) [inline src/main/kotlin/org/jetbrains/research/ml/ast/gumtree/diff/PsiElementTransformer.kt:20]: Why do we need to pass srcPsiNodes, dstPsiNodes, and insertedNodesIds to the constructor? what if we pass the wrong variables? why not make fields and count automatically?

**Review by elena-lyulina — CHANGES_REQUESTED** [review]
Just one more comment

**Review by elena-lyulina — APPROVED** [review]
Nice :)

**Pull request merged**